### ISSUE TITLE
Add a way to override the resultant Python package version

### ISSUE THREAD

**Issue by DavidAntliff** [PyO3/maturin#2163]
Currently, `maturin build` uses the `project.version` field from `Cargo.toml` to version the resultant Python package.

Using this version is not recommended by the Cargo team, as it has a specific crates.io purpose, and allegedly should not be overloaded as a general "version" number for a project: https://github.com/rust-lang/cargo/issues/6583

I suggest a feature to enable override of the version number that `maturin build` (and `maturin develop`) assigns to the resultant Python package, so that alternative versioning schemes can be used.

As of today, the only way I can find to influence the Python package version seems to be to modify `project.version` prior to build, such as via `sed` in a CI script, but this does not work well for local workflows such as developer builds where a CI version number is not available. It also marks `Cargo.toml` as modified which can have side-effects for source control. Modifying source files is not considered proper in reproducible build environments like "yocto" or Debian package system, either, where upstream source files need to be considered authoritative.

So I suggest that there could be a new way for the _environment_ (e.g. shell environment, or script that invokes `maturin`) to modify the resultant version number without modifying source files.

**Comment by ia0** [comment]
Note also that if `cargo build` reads its own version (which is the case when writing a CLI with `clap` for the `--version` flag), then the workaround of editing the `Cargo.toml` doesn't work, because we want the crate version for the CLI but the Python library version for the Python package.

**Comment by konstin** [comment]
Why don't you use `project.version` from `pyproject.toml` instead, which is the dedicated field for this and has the advantage of also providing static metadata to Python tooling?

**Comment by reyammer** [comment]
For magika we specify the version in `magika.__init__` [here](https://github.com/google/magika/blob/9ff1678b70e3a9a03f2f9dacf527a4ca06ac4a3e/python/src/magika/__init__.py#L18) as we want to allow version introspection with `magika.__version__`. And to avoid duplicating info, we used the 'dynamic version' feature in pyproject [here](https://github.com/google/magika/blob/9ff1678b70e3a9a03f2f9dacf527a4ca06ac4a3e/python/pyproject.toml#L34).

At that time, this seemed the best practice to achieve this and having one single source of truth.

Maybe the modern way is to not do that, and use something like `importlib.metadata.version`? What would be your recommendation?

Thanks!

**Comment by messense** [comment]
Could you set the `__version__` in Rust side like `m.add("__version__", env!("CARGO_PKG_VERSION"))?;` then re-export it at `magika.__version__`?

**Comment by konstin** [comment]
> Maybe the modern way is to not do that, and use something like `importlib.metadata.version`? What would be your recommendation?

Yes! You can set `__version__` to the output of `importlib.metadata.version` for easier migration.

**Comment by ia0** [comment]
> Could you set the `__version__` in Rust side like `m.add("__version__", env!("CARGO_PKG_VERSION"))?;` then re-export it at `magika.__version__`?

For this particular use-case (Magika), Rust is just a binary, not a library.

**Comment by reyammer** [comment]
Thank you @konstin, this seems to have worked (and we could get rid of our hack to patch the Cargo version with the python package version).

Commit for whoever steps by this same problem: https://github.com/google/magika/pull/1265/changes/b5c9ff3020a599d5031dca89ce2939533190ca0d

Thanks again!

**Comment by antonagestam** [comment]
I wonder if it would be feasible for maturin to support setuptools-scm. I'm migrating a project currently using that and it would be really convenient not to have to go back to hard-coding versions again.

**Comment by antonagestam** [comment]
Perhaps it's worth considering offering a similar plugin hook as to how setuptools solves this, with a `finalize_distribution_options` entry-point.

- https://setuptools.pypa.io/en/latest/userguide/extension.html#customizing-distribution-options
- https://github.com/pypa/setuptools-scm/blob/b599da76c633784a9514a7e62aba98ca7cc7c3c6/pyproject.toml#L95
